src/erlcloud_ssm.erl

```diff
@@ -0,0 +1,589 @@
+       ssm_host=Host,
+       ssm_port=Port
+      }.
+
+-spec configure(string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey)),
+    ok.
+
+-spec configure(string(), string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host, Port) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port)),
+    ok.
+
+default_config() ->
+    erlcloud_aws:default_config().
+
+
+%%%------------------------------------------------------------------------------
+%%% Shared types
+%%%------------------------------------------------------------------------------
+
+-type string_param() :: binary() | string().
+
+-type json_pair() :: {binary() | atom(), jsx:json_term()}.
+-type json_return() :: {ok, jsx:json_term()} | {error, term()}.
+-type ssm_return(Record) :: {ok, jsx:json_term() | Record } | {error, term()}.
+-type decode_fun() :: fun((jsx:json_term(), decode_opts()) -> tuple()).
+
+%%%------------------------------------------------------------------------------
+%%% Shared Options
+%%%------------------------------------------------------------------------------
+-type out_type() :: json | record | map.
+-type out_opt() :: {out, out_type()}.
+-type property() :: proplists:property().
+
+-type aws_opts() :: [json_pair()].
+-type ssm_opts() :: [out_opt()].
+-type opts() :: {aws_opts(), ssm_opts()}.
+
+-spec verify_ssm_opt(atom(), term()) -> ok.
+verify_ssm_opt(out, Value) ->
+    case lists:member(Value, ?OUTPUT_CHOICES) of
+        true ->
+            ok;
+        false ->
+            error({erlcloud_ssm, {invalid_opt, {out, Value}}})
+    end;
+verify_ssm_opt(Name, Value) ->
+    error({erlcloud_ssm, {invalid_opt, {Name, Value}}}).
+
+-type opt_table_entry() :: {atom(), binary(), fun((_) -> jsx:json_term())}.
+-type opt_table() :: [opt_table_entry()].
+-spec opt_folder(opt_table(), property(), opts()) -> opts().
+opt_folder(_, {_, undefined}, Opts) ->
+    %% ignore options set to undefined
+    Opts;
+opt_folder(Table, {Name, Value}, {AwsOpts, EcsOpts}) ->
+    case lists:keyfind(Name, 1, Table) of
+        {Name, Key, ValueFun} ->
+            {[{Key, ValueFun(Value)} | AwsOpts], EcsOpts};
+        false ->
+            verify_ssm_opt(Name, Value),
+            {AwsOpts, [{Name, Value} | EcsOpts]}
+    end.
+
+-spec opts(opt_table(), proplist()) -> opts().
+opts(Table, Opts) when is_list(Opts) ->
+    %% remove duplicate options
+    Opts1 = lists:ukeysort(1, proplists:unfold(Opts)),
+    lists:foldl(fun(Opt, A) -> opt_folder(Table, Opt, A) end, {[], []}, Opts1);
+opts(_, _) ->
+    error({erlcloud_ssm, opts_not_list}).
+
+%%%------------------------------------------------------------------------------
+%%% Shared Decoders
```

Comment:
Any reason why this one is 3 `%` and not 2 `%` like most other spacers in the module? There are several in here like this.